### PR TITLE
[Mission] CoP 6-4 Mechanics

### PR DESCRIPTION
**Pull request by The-Aerec**
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)
- [x]  I've _**tested my code and the things my code has changed**_ since the last commit in the PR, and will test after any later commits

## What does this pull request do?

Adds Mechanics for Mammets including Skill changes on Weapon Change
Adds Mechanics for Omega
Adds Mechanics for Ultima

## Steps to test these changes

!togglegm
Participate in the BCNM for the mission 'One to be Feared'

### CODE REVIEW

File: src/map/entities/battleentity.h [hunk 1]
```diff
@@ -423,6 +423,9 @@ enum IMMUNITY : uint16
     IMMUNITY_REQUIEM     = 0x400,  // 1024
     IMMUNITY_LIGHT_SLEEP = 0x800,  // 2048
     IMMUNITY_DARK_SLEEP  = 0x1000, // 4096
+    IMMUNITY_ASPIR = 1 << 13,     // 8192
+    IMMUNITY_TERROR = 1 << 14,    // 16384
```
Comment (The-Aerec) [inline src/map/entities/battleentity.h]: Still in progress, on Wings base it was as simple as adding these additional immunity enums for needing new resists, but maybe some other work still needs to be done here. 

Jettatura (Terror) should return immune on omega / ultima

Comment (The-Aerec) [reply src/map/entities/battleentity.h]: Fixed.

File: src/map/status_effect_container.cpp [hunk 2]
```diff
@@ -349,6 +349,18 @@ bool CStatusEffectContainer::CanGainStatusEffect(CStatusEffect* PStatusEffect)
...
+            {
+                return false;
+            }
+            break;
+        case EFFECT_TERROR:
+            if (m_POwner->hasImmunity(IMMUNITY_TERROR))
+            {
+                return false;
+            }
+            break;
```
Comment (Xaver-DaRed) [inline src/map/status_effect_container.cpp]:
```
        case EFFECT_TERROR:
            if (m_POwner->hasImmunity(IMMUNITY_TERROR))
            {
                return false;
            }
            break;
        case EFFECT_TERROR:
            if (m_POwner->hasImmunity(IMMUNITY_TERROR))
            {
                return false;
            }
            break;
```

Duplicated?

**Comment by Xaver-DaRed** [comment]
Maybe we could move the form code of mammets to a mixin

File: scripts/zones/Sealions_Den/mobs/Omega.lua [hunk 3]
```diff
@@ -3,6 +3,7 @@
 --  Mob: Omega
 -----------------------------------
 local oneToBeFeared = require("scripts/zones/Sealions_Den/helpers/One_to_be_Feared")
+local ID = require("scripts/zones/Sealions_Den/IDs")
```
Comment (Xaver-DaRed) [inline scripts/zones/Sealions_Den/mobs/Omega.lua]: I don't think this require is needed

File: src/map/status_effect_container.cpp (hunk 2 above)
Comment (The-Aerec) [reply src/map/status_effect_container.cpp]: Apparently so? I must have copy pasted accidentally before I committed, I had just built it before..

**Comment by claywar** [comment]
Just a heads up, Co-authored-by lines need to be in the message of the commit, not the title.  Also, there is a required format for it.

File: scripts/globals/mobskills/target_analysis.lua [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+---------------------------------------------
+require("scripts/globals/mobskills")
+require("scripts/globals/status")
+require("scripts/settings/main")
+local ID = require("scripts/zones/Sealions_Den/IDs")
+local ID2 = require("scripts/zones/Apollyon/IDs")
+---------------------------------------------
+local mobskill_object = {}
+
+local attributes =
```
Comment (claywar) [inline scripts/globals/mobskills/target_analysis.lua]: This table name should probably be more specific.  Its not a table of attributes, but attribute down effects.

File: scripts/globals/mobskills/target_analysis.lua [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+            ID.mob.OMEGA_OFFSET + i == id or
+            ID2.mob.APOLLYON_CS_MOB == id
+        then
+             checkID = true
+        end
+    end
+
+    if checkID and
+       mobhp > 25 or
+       checkID and
```
Comment (claywar) [inline scripts/globals/mobskills/target_analysis.lua]: Would suggest parens to make logic less ambiguous

File: scripts/globals/mobskills/target_analysis.lua [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    xi.effect.AGI_DOWN,
+    xi.effect.MND_DOWN,
+    xi.effect.INT_DOWN,
+    xi.effect.CHR_DOWN,
+}
+
+mobskill_object.onMobSkillCheck = function(target, mob, skill)
+    local mobhp = mob:getHPP()
+    local id = mob:getID()
+    local checkID = false
```
Comment (claywar) [inline scripts/globals/mobskills/target_analysis.lua]: This seems a bit unclear (and looking at the loop), what is the purpose of this?

File: scripts/zones/Sealions_Den/mobs/Mammet-22_Zeta.lua [hunk 7]
```diff
@@ -20,21 +19,39 @@ local forms =
     STAFF   = 3,
 }
 
+local tpMoves =
```
Comment (claywar) [inline scripts/zones/Sealions_Den/mobs/Mammet-22_Zeta.lua:22]: For this and the table above, would suggest breaking this out to single items per line, and commenting there

File: src/map/entities/battleentity.h [hunk 8]
```diff
@@ -423,6 +423,9 @@ enum IMMUNITY : uint16
     IMMUNITY_REQUIEM     = 0x400,  // 1024
     IMMUNITY_LIGHT_SLEEP = 0x800,  // 2048
     IMMUNITY_DARK_SLEEP  = 0x1000, // 4096
+    IMMUNITY_ASPIR = 1 << 13,     // 8192
```
Comment (claywar) [inline src/map/entities/battleentity.h]: Should follow the previously used styling for this.  While I'm a fan of declarations using bitshifts, it doesn't follow current formatting

File: scripts/globals/mobskills/target_analysis.lua [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    xi.effect.MND_DOWN,
+    xi.effect.INT_DOWN,
+    xi.effect.CHR_DOWN,
+}
+
+mobskill_object.onMobSkillCheck = function(target, mob, skill)
+    local skillList = mob:getMobMod(xi.mobMod.SKILL_LIST)
+    local mobhp = mob:getHPP()
+
+    if ((skillList == 54 and mobhp < 26) or (skillList == 727 and mob:getAnimationSub() == 1)) then
```
Comment (claywar) [inline scripts/globals/mobskills/target_analysis.lua]: Unnecessary outer parens can be removed here

File: scripts/globals/mobskills/target_analysis.lua [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    else
+        return 1
+    end
+end
+
+mobskill_object.onMobWeaponSkill = function(target, mob, skill)
+    local drained = 0
+
+    for i = 1,7 do
+        if math.random(0,100) < 40 then
```
Comment (claywar) [inline scripts/globals/mobskills/target_analysis.lua]: Space between parameters please

File: scripts/globals/mobskills/transmogrification.lua [hunk 11]
```diff
@@ -12,7 +12,7 @@ require("scripts/globals/msg")
 local mobskill_object = {}
 
 mobskill_object.onMobSkillCheck = function(target, mob, skill)
-    if (mob:getAnimationSub() == 3) then
+    if (mob:getAnimationSub() == 0) then
```
Comment (claywar) [inline scripts/globals/mobskills/transmogrification.lua]: Outer parens can be removed

File: scripts/zones/Monarch_Linn/mobs/Mammet-19_Epsilon.lua [hunk 12]
```diff
@@ -6,49 +6,97 @@ require("scripts/globals/status")
...
-    -- Mammets seem to be able to change to any given form, per YouTube videos
-    -- Added a random chance to change forms every 3 seconds if 60 seconds have passed, just to make things less formulaic.
-        -- May be able to change forms more often.  Witnessed one at ~50 seconds, most were 60-80.
-        -- Believe it or not, these changes may be too slow @ 50% chance.  Probability is a pain.
-    -- L40 means their "weapons" are 40 DMG by default.
-    if ((mob:getBattleTime() > mob:getLocalVar('changeTime') + 60 or mob:getLocalVar('changeTime') == 0) and math.random(0, 1) == 1
-        and not mob:hasStatusEffect(xi.effect.FOOD)) then
-        changeForm(mob)
+    -- Chages forms after 30-60 seconds randomly
+    local TT = mob:getLocalVar("formTimeTracker")
```
Comment (claywar) [inline scripts/zones/Monarch_Linn/mobs/Mammet-19_Epsilon.lua]: Upper-case starting localvar needs to be renamed

File: scripts/zones/Sealions_Den/mobs/Mammet-22_Zeta.lua [hunk 13]
```diff
@@ -20,21 +19,46 @@ local forms =
...
 end
 
 entity.onMobFight = function(mob, target)
-    -- TODO: What are the conditions for a Mammet to change forms?
-    --       For now we'll change forms every 20% health
-    local healthTracker = mob:getLocalVar("formHealthTracker")
-    local currentHealth = mob:getHPP()
-
+    -- Chages forms after 30-60 seconds randomly
+    local TT = mob:getLocalVar("formTimeTracker")
```
Comment (claywar) [inline scripts/zones/Sealions_Den/mobs/Mammet-22_Zeta.lua]: Same as above

File: src/map/entities/battleentity.h [hunk 14]
```diff
@@ -423,7 +423,9 @@ enum IMMUNITY : uint16
     IMMUNITY_REQUIEM     = 0x400,  // 1024
     IMMUNITY_LIGHT_SLEEP = 0x800,  // 2048
     IMMUNITY_DARK_SLEEP  = 0x1000, // 4096
-};
+    IMMUNITY_ASPIR       = 0x2000, // 8192
+    IMMUNITY_TERROR      = 0x4000, // 16384
+    IMMUNITY_DISPEL      = 0x8000, // 32768
```
Comment (claywar) [inline src/map/entities/battleentity.h:430]: No closing brace here after removal

File: src/map/status_effect_container.cpp [hunk 15]
```diff
@@ -349,6 +349,12 @@ bool CStatusEffectContainer::CanGainStatusEffect(CStatusEffect* PStatusEffect)
                 return false;
             }
             break;
+        case EFFECT_TERROR:
```
Comment (claywar) [inline src/map/status_effect_container.cpp:352]: Fairly certain this was added in a previous PR

Comment (The-Aerec) [reply src/map/status_effect_container.cpp:352]: a previous one of mine? I had to add this here to get the terror immunity to work. If there is a better way let me know

Comment (claywar) [reply src/map/status_effect_container.cpp:352]: Correct, pretty sure I saw this in a recently-merged PR (or one that's still open)

Comment (The-Aerec) [reply src/map/status_effect_container.cpp:352]: Just rechecked, this is the only one i've made core changes in.

**Review by claywar — APPROVED** [review]

**Pull request merged**